### ISSUE TITLE
cluster-kube-apiserver-operator -- unknown flag: --infra-config-file

### ISSUE THREAD

**Issue by chtime** [openshift/installer#5366]
# Version

```console
$ openshift-install version
./openshift-install unreleased-master-4696-g1c11fbcac64b85985ed1c30852317164cab06f10
built from commit 1c11fbcac64b85985ed1c30852317164cab06f10
release image registry.ci.openshift.org/origin/release:4.8
```

# Platform:
baremetal UPI

# What happened?

- Created a basic `install-config.yaml` based on [the doc](https://github.com/openshift/installer/blob/release-4.8/docs/user/metal/install_upi.md#setting-up-install-config-for-installer)
- `openshift-install create ignition-configs`
- start a virtual machine running any of the following
  - rhcos48
  - rhcos410
  - fedora coreos 34
- point `core-installer` to the `bootstrap.ign` created before
- reboot the bootstrap node
- the bootstrap node hangs when rendering the api server manifests:

```
Nov 10 09:29:55 oc-bootstrap bootkube.sh[1743]: Rendering Kubernetes API server core manifests...
Nov 10 09:29:57 oc-bootstrap podman[5493]: 2021-11-10 09:29:57.545982311 +0000 UTC m=+2.850595529 container cleanup 2a98b9caeb6a25dee4c200d5cf4f6a25dc6dc3c9d52639789b97a54ec28329ca (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:0ea51558c0234d7c21b10a598f56d2ee20b9bdf4bc229db227b5938e43a94388, name=config-render, io.openshift.build.commit.ref=master, io.openshift.release.operator=true, io.buildah.version=1.20.1, io.openshift.build.commit.url=https://github.com/openshift/ocp-build-data/commit/f17f5523faca5bd928d5cc65484a66b20f04f366, io.openshift.build.source-context-dir=, release=202106011824.p0.git.f17f552.assembly.stream, vendor=Red Hat, Inc., io.openshift.ci.from.base=sha256:38a303528a5de6cbdd5805f306002dcde4f5e37c5b6caa8ce4393717c3161dc3, io.k8s.description=This is the base image from which all OpenShift images inherit., io.openshift.build.namespace=, distribution-scope=public, io.openshift.maintainer.product=OpenShift Container Platform, name=openshift/base-rhel8, io.openshift.expose-services=, com.redhat.license_terms=https://www.redhat.com/agreements, io.openshift.build.source-location=https://github.com/openshift/cluster-config-operator, vcs-url=https://github.com/openshift/cluster-config-operator, io.k8s.display-name=OpenShift Base, build-date=2021-06-01T19:07:35.203452, vcs-ref=c1022410bf9a000e84d9764f8e8a5cfa35cfa452, io.openshift.build.commit.date=, com.redhat.component=openshift-base-rhel8-container, io.openshift.maintainer.component=Release, io.openshift.build.commit.message=, com.redhat.build-host=cpt-1008.osbs.prod.upshift.rdu2.redhat.com, description=The Universal Base Image is designed and engineered to be the base layer for all of your containerized applications, middleware and utilities. This base image is freely redistributable, but Red Hat only supports Red Hat technologies through subscriptions for Red Hat products. This image is maintained by Red Hat and updated regularly., io.openshift.tags=base rhel8, maintainer=Red Hat, Inc., vcs-type=git, io.openshift.build.commit.id=c1022410bf9a000e84d9764f8e8a5cfa35cfa452, io.openshift.build.commit.author=, summary=Provides the latest release of Red Hat Universal Base Image 8., architecture=x86_64, version=v4.8.0, io.openshift.build.name=, url=https://access.redhat.com/containers/#/registry.access.redhat.com/openshift/base-rhel8/images/v4.8.0-202106011824.p0.git.f17f552.assembly.stream)
Nov 10 09:30:06 oc-bootstrap podman[5578]: 2021-11-10 09:29:55.211440779 +0000 UTC m=+0.067921954 image pull  registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184
Nov 10 09:30:06 oc-bootstrap podman[5578]: 2021-11-10 09:30:06.968503735 +0000 UTC m=+11.824984354 container create b7f435bf5394e29b02d0780661485801243f8e8342549a82b2e3c220a308713c (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184, name=kube-apiserver-render, vcs-url=https://github.com/openshift/cluster-kube-apiserver-operator, name=openshift/ose-base, io.openshift.build.source-location=https://github.com/openshift/cluster-kube-apiserver-operator, summary=Provides the latest release of Red Hat Universal Base Image 8., vendor=Red Hat, Inc., url=https://access.redhat.com/containers/#/registry.access.redhat.com/openshift/ose-base/images/v4.8.0-202110262219.p0.git.ad38e11.assembly.stream, io.openshift.build.source-context-dir=, io.openshift.maintainer.product=OpenShift Container Platform, io.openshift.release.operator=true, License=GPLv2+, io.openshift.tags=openshift,base, build-date=2021-10-26T23:26:35.287089, io.buildah.version=1.20.1, distribution-scope=public, vcs-ref=cddab90b7699ceb50032e22b4b3cff5803dd1466, io.k8s.description=ART equivalent image openshift-4.8-openshift-enterprise-base - rhel-8/base-repos, description=This is the base image from which all OpenShift Container Platform images inherit., vcs-type=git, com.redhat.license_terms=https://www.redhat.com/agreements, io.openshift.build.commit.date=, io.openshift.build.name=, architecture=x86_64, io.openshift.build.commit.author=, io.openshift.build.commit.id=cddab90b7699ceb50032e22b4b3cff5803dd1466, io.openshift.build.namespace=, io.openshift.build.commit.url=https://github.com/openshift/images/commit/ad38e116f3533e0e78621a2887d8b70cde7cba31, io.k8s.display-name=4.8-base, io.openshift.build.commit.message=, io.openshift.ci.from.base=sha256:da8d56b1cada330756f42f5852017cafbd726c12d66e1b531def067fd18355e4, version=v4.8.0, com.redhat.build-host=cpt-1007.osbs.prod.upshift.rdu2.redhat.com, maintainer=Red Hat, Inc., io.openshift.build.commit.ref=release-4.8, io.openshift.expose-services=, release=202110262219.p0.git.ad38e11.assembly.stream, com.redhat.component=openshift-enterprise-base-container, io.openshift.maintainer.component=Release)
Nov 10 09:30:07 oc-bootstrap podman[5578]: 2021-11-10 09:30:07.57977905 +0000 UTC m=+12.436259916 container init b7f435bf5394e29b02d0780661485801243f8e8342549a82b2e3c220a308713c (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184, name=kube-apiserver-render, io.openshift.build.commit.message=, summary=Provides the latest release of Red Hat Universal Base Image 8., io.openshift.ci.from.base=sha256:da8d56b1cada330756f42f5852017cafbd726c12d66e1b531def067fd18355e4, vcs-url=https://github.com/openshift/cluster-kube-apiserver-operator, distribution-scope=public, release=202110262219.p0.git.ad38e11.assembly.stream, License=GPLv2+, io.openshift.expose-services=, description=This is the base image from which all OpenShift Container Platform images inherit., io.openshift.build.commit.author=, name=openshift/ose-base, io.openshift.build.name=, io.openshift.build.commit.ref=release-4.8, url=https://access.redhat.com/containers/#/registry.access.redhat.com/openshift/ose-base/images/v4.8.0-202110262219.p0.git.ad38e11.assembly.stream, io.openshift.build.namespace=, io.k8s.display-name=4.8-base, maintainer=Red Hat, Inc., version=v4.8.0, io.buildah.version=1.20.1, architecture=x86_64, io.openshift.build.commit.date=, io.openshift.build.commit.url=https://github.com/openshift/images/commit/ad38e116f3533e0e78621a2887d8b70cde7cba31, io.k8s.description=ART equivalent image openshift-4.8-openshift-enterprise-base - rhel-8/base-repos, vcs-type=git, com.redhat.component=openshift-enterprise-base-container, io.openshift.tags=openshift,base, vendor=Red Hat, Inc., com.redhat.license_terms=https://www.redhat.com/agreements, build-date=2021-10-26T23:26:35.287089, io.openshift.build.commit.id=cddab90b7699ceb50032e22b4b3cff5803dd1466, io.openshift.build.source-context-dir=, io.openshift.maintainer.component=Release, io.openshift.release.operator=true, io.openshift.build.source-location=https://github.com/openshift/cluster-kube-apiserver-operator, com.redhat.build-host=cpt-1007.osbs.prod.upshift.rdu2.redhat.com, io.openshift.maintainer.product=OpenShift Container Platform, vcs-ref=cddab90b7699ceb50032e22b4b3cff5803dd1466)
Nov 10 09:30:07 oc-bootstrap podman[5578]: 2021-11-10 09:30:07.63525139 +0000 UTC m=+12.491732013 container start b7f435bf5394e29b02d0780661485801243f8e8342549a82b2e3c220a308713c (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184, name=kube-apiserver-render, vendor=Red Hat, Inc., com.redhat.license_terms=https://www.redhat.com/agreements, release=202110262219.p0.git.ad38e11.assembly.stream, io.k8s.description=ART equivalent image openshift-4.8-openshift-enterprise-base - rhel-8/base-repos, io.openshift.build.commit.author=, com.redhat.build-host=cpt-1007.osbs.prod.upshift.rdu2.redhat.com, io.k8s.display-name=4.8-base, io.openshift.build.commit.message=, vcs-ref=cddab90b7699ceb50032e22b4b3cff5803dd1466, vcs-url=https://github.com/openshift/cluster-kube-apiserver-operator, name=openshift/ose-base, io.openshift.build.source-location=https://github.com/openshift/cluster-kube-apiserver-operator, distribution-scope=public, description=This is the base image from which all OpenShift Container Platform images inherit., summary=Provides the latest release of Red Hat Universal Base Image 8., io.openshift.maintainer.component=Release, io.openshift.build.commit.id=cddab90b7699ceb50032e22b4b3cff5803dd1466, url=https://access.redhat.com/containers/#/registry.access.redhat.com/openshift/ose-base/images/v4.8.0-202110262219.p0.git.ad38e11.assembly.stream, io.openshift.build.namespace=, io.openshift.build.name=, License=GPLv2+, io.openshift.tags=openshift,base, version=v4.8.0, io.openshift.release.operator=true, io.openshift.build.commit.ref=release-4.8, io.openshift.ci.from.base=sha256:da8d56b1cada330756f42f5852017cafbd726c12d66e1b531def067fd18355e4, com.redhat.component=openshift-enterprise-base-container, maintainer=Red Hat, Inc., build-date=2021-10-26T23:26:35.287089, io.openshift.build.source-context-dir=, io.openshift.build.commit.url=https://github.com/openshift/images/commit/ad38e116f3533e0e78621a2887d8b70cde7cba31, vcs-type=git, io.openshift.expose-services=, io.buildah.version=1.20.1, io.openshift.maintainer.product=OpenShift Container Platform, io.openshift.build.commit.date=, architecture=x86_64)
Nov 10 09:30:07 oc-bootstrap podman[5578]: 2021-11-10 09:30:07.635558212 +0000 UTC m=+12.492038737 container attach b7f435bf5394e29b02d0780661485801243f8e8342549a82b2e3c220a308713c (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184, name=kube-apiserver-render, version=v4.8.0, io.k8s.display-name=4.8-base, distribution-scope=public, io.openshift.build.commit.ref=release-4.8, vcs-ref=cddab90b7699ceb50032e22b4b3cff5803dd1466, io.openshift.build.namespace=, io.openshift.build.name=, io.openshift.maintainer.product=OpenShift Container Platform, vendor=Red Hat, Inc., io.openshift.build.commit.date=, description=This is the base image from which all OpenShift Container Platform images inherit., name=openshift/ose-base, com.redhat.build-host=cpt-1007.osbs.prod.upshift.rdu2.redhat.com, maintainer=Red Hat, Inc., summary=Provides the latest release of Red Hat Universal Base Image 8., io.openshift.build.commit.message=, io.openshift.ci.from.base=sha256:da8d56b1cada330756f42f5852017cafbd726c12d66e1b531def067fd18355e4, io.openshift.build.source-context-dir=, vcs-url=https://github.com/openshift/cluster-kube-apiserver-operator, io.openshift.release.operator=true, io.openshift.build.source-location=https://github.com/openshift/cluster-kube-apiserver-operator, License=GPLv2+, io.openshift.maintainer.component=Release, build-date=2021-10-26T23:26:35.287089, io.openshift.tags=openshift,base, io.openshift.build.commit.author=, vcs-type=git, com.redhat.license_terms=https://www.redhat.com/agreements, com.redhat.component=openshift-enterprise-base-container, architecture=x86_64, io.openshift.build.commit.id=cddab90b7699ceb50032e22b4b3cff5803dd1466, io.openshift.expose-services=, release=202110262219.p0.git.ad38e11.assembly.stream, io.openshift.build.commit.url=https://github.com/openshift/images/commit/ad38e116f3533e0e78621a2887d8b70cde7cba31, io.buildah.version=1.20.1, url=https://access.redhat.com/containers/#/registry.access.redhat.com/openshift/ose-base/images/v4.8.0-202110262219.p0.git.ad38e11.assembly.stream, io.k8s.description=ART equivalent image openshift-4.8-openshift-enterprise-base - rhel-8/base-repos)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]: Error: unknown flag: --infra-config-file
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]: Usage:
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:   cluster-kube-apiserver-operator render [flags]
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]: Flags:
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --asset-input-dir string                     A path to directory with certificates and secrets.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --asset-output-dir string                    Output path for rendered manifests.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --cluster-auth-file string                   Openshift Cluster Authentication API Config file.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --cluster-config-file string                 Openshift Cluster API Config file.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --config-output-file string                  Output path for the KubeAPIServerConfig.kubecontrolplane.config.openshift.io/v1 yaml file.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --config-override-files strings              Additional sparse KubeAPIServerConfig.kubecontrolplane.config.openshift.io/v1 files for customiziation through the installer, merged into the default config in the given order.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:   -h, --help                                       help for render
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-cloud-provider-host-path string   A host path mounted into the apiserver pods to hold cloud provider configuration. (default "/etc/kubernetes/cloud")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-config-file-name string           The config file name inside the manifest-config-host-path. (default "kube-apiserver-config.yaml")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-config-host-path string           A host path mounted into the apiserver pods to hold a config file. (default "/etc/kubernetes/bootstrap-configs")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-etcd-server-urls stringArray      The etcd server URL, comma separated. (default [https://127.0.0.1:2379])
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-etcd-serving-ca string            The etcd serving CA. (default "root-ca.crt")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-image string                      Image to use for the apiserver. (default "openshift/origin-hyperkube:latest")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-image-pull-policy string          Image pull policy to use for the apiserver. (default "IfNotPresent")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-lock-host-path string             A host path mounted into the apiserver pods to hold lock. (default "/var/run/kubernetes/lock")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-namespace string                  Target namespace for phase 3 apiserver pods. (default "openshift-kube-apiserver")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-operator-image string             Operator image to use for the apiserver.
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --manifest-secrets-host-path string          A host path mounted into the apiserver pods to hold secrets. (default "/etc/kubernetes/bootstrap-secrets")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --templates-input-dir string                 A path to a directory with manifest templates. (default "/usr/share/bootkube/manifests")
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]: Global Flags:
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --add-dir-header                   If true, adds the file directory to the header of the log messages
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --alsologtostderr                  log to standard error as well as files
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --log-backtrace-at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --log-dir string                   If non-empty, write log files in this directory
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --log-file string                  If non-empty, use this log file
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --log-flush-frequency duration     Maximum number of seconds between log flushes (default 5s)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --logtostderr                      log to standard error instead of files (default true)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --skip-headers                     If true, avoid header prefixes in the log messages
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --skip-log-headers                 If true, avoid headers when opening log files
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:   -v, --v Level                          number for the log level verbosity
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]:       --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
Nov 10 09:30:07 oc-bootstrap bootkube.sh[5578]: unknown flag: --infra-config-file
Nov 10 09:30:07 oc-bootstrap podman[5578]: 2021-11-10 09:30:07.726236161 +0000 UTC m=+12.582716788 container died b7f435bf5394e29b02d0780661485801243f8e8342549a82b2e3c220a308713c (image=registry.ci.openshift.org/origin/4.8-2021-11-09-000235@sha256:115df224b9027438331c475e147ded4f5004dc16aa8803517fd357a61ea9d184, name=kube-apiserver-render)
Nov 10 09:30:07 oc-bootstrap systemd[1]: bootkube.service: Main process exited, code=exited, status=1/FAILURE
Nov 10 09:30:07 oc-bootstrap systemd[1]: bootkube.service: Failed with result 'exit-code'.
```

# What you expected to happen?

I expected to have a running bootstrap instance providing a kubeapi at 6443 and a machine server at 22623.

# How to reproduce it (as minimally and precisely as possible)?

See above. I tried with installer branches  master, release-4.10, release-4.9 and release-4.8, and the bootstrap node running rhcos48, rhcos410 and fedora coreos 34; the latter indicated in the podman logs that the bootkube job pulled container versioned 4.8, which is why I ultimately thought that installer version 4.8 should work and opened this issue mentioning this version.

# Anything else we need to know?

The target system is a KVM hypervisor which is eventually supposed to be running 5 VM's: bootstrap (deleted eventually), control{1-3} and a single worker, worker1. This setup is obviously not for proper production purposes, but for learning and evaluation.

I suspect that there is another version flag that I have to adapt to get the proper container version to be pulled.

For reference, this is the `install-config.yaml`:

```yaml
apiVersion: v1
baseDomain: <redacted>
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform: {}
  replicas: 1
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform: {}
  replicas: 1
metadata:
  creationTimestamp: null
  name: oc
networking:
  clusterNetwork:
  - cidr: 10.11.0.0/16
    hostPrefix: 23
  machineNetwork:
  - cidr: 10.10.0.0/24
  networkType: OpenShiftSDN
  serviceNetwork:
  - 172.30.0.0/16
platform:
  none: {}
publish: External
pullSecret: <redacted>
sshKey: <redacted>
```

# References

- #5081 was closed because reverting to 4.8 seemed to have worked there, but it doesn't for me.

**Comment by staebler** [comment]
Try building branch `release-4.9` and setting the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` environment variable to the `pullSpec` field from https://openshift-release.apps.ci.l2s4.p1.openshiftapps.com/api/v1/releasestream/4-stable/latest.

**Comment by chtime** [comment]
I checked out `release-4.9`, set the override (`export OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE='quay.io/openshift-release-dev/ocp-release:4.9.8-x86_64'`), rebuilt (`./hack/build.sh`), re-generated ignition files and `coreos-installer install`-ed the bootstrap node with the new ignition files and exactly the same thing happened. :-/

Should the coreos image I am installing on the bootstrap node be a very specific one? I tried again with rhcos-48 and fedora's coreos-34. Is there any other infos I can provide?
